geniza/corpus/management/commands/escr_alto_to_annotation.py

```diff
@@ -0,0 +1,249 @@
+import re
+
+from django.conf import settings
+from django.contrib.admin.models import ADDITION, LogEntry
+from django.contrib.auth.models import User
+from django.contrib.contenttypes.models import ContentType
+from django.core.management.base import BaseCommand
+from django.db.models import Q
+from djiffy.models import Canvas, Manifest
+from eulxml import xmlmap
+
+from geniza.annotations.models import Annotation
+from geniza.corpus.models import Document
+from geniza.footnotes.models import Footnote, Source, SourceType
+
+# parse eScriptorium ALTO files, generate transcription IIIF annotations,
+# and attach to correct documents
+
+
+class AltoObject(xmlmap.XmlObject):
+    # all XmlObjects should inherit this class to get the root namespace
+    ROOT_NAMESPACES = {"alto": "http://www.loc.gov/standards/alto/ns-v4#"}
+
+
+class AltoPolygonalObject(AltoObject):
+    polygon = xmlmap.NodeField("alto:Shape/alto:Polygon/@POINTS", AltoObject)
+
+
+class Line(AltoPolygonalObject):
+    content = xmlmap.StringField("alto:String/@CONTENT")
+
+
+class TextBlock(AltoPolygonalObject):
+    id = xmlmap.StringField("./@ID")
+    lines = xmlmap.NodeListField("alto:TextLine", Line)
+    block_type_id = xmlmap.StringField("./@TAGREFS")
+
+
+class PrintSpace(AltoObject):
+    textblocks = xmlmap.NodeListField("alto:TextBlock", TextBlock)
+
+
+class EscriptoriumAlto(AltoObject):
+    filename = xmlmap.StringField(
+        "//alto:alto/alto:Description/alto:sourceImageInformation/alto:fileName"
+    )
+    printspace = xmlmap.NodeField(
+        "//alto:alto/alto:Layout/alto:Page/alto:PrintSpace", PrintSpace
+    )
+    tags = xmlmap.NodeListField("//alto:alto/alto:Tags/alto:OtherTag", AltoObject)
+
+
+class Command(BaseCommand):
+    # regex pattern for image filenames
+    filename_pattern = r"PGPID_(?P<pgpid>\d+)_(?P<shelfmark>[\w\-]+)_(?P<img>\d)\..+"
+
+    def add_arguments(self, parser):
+        # needs xml filenames as input
+        parser.add_argument(
+            "alto", metavar="ALTOXML", nargs="+", help="ALTO files to be processed"
+        )
+
+    def handle(self, *args, **options):
+        self.script_user = User.objects.get(username=settings.SCRIPT_USERNAME)
+
+        for xmlfile in options["alto"]:
+            self.stdout.write("Processing %s" % xmlfile)
+            self.ingest_xml(xmlfile)
+
+    def ingest_xml(self, xmlfile):
+        alto = xmlmap.load_xmlobject_from_file(xmlfile, EscriptoriumAlto)
+        # associate filename with pgpid
+        m = re.match(self.filename_pattern, alto.filename)
+        pgpid = int(m.group("pgpid"))
+        try:
+            doc = Document.objects.get(Q(id=pgpid) | Q(old_pgpids__contains=[pgpid]))
```

Comment:
Is this a common enough operation that it would be worth a custom method on `DocumentQuerySet` ? I know we do something similar in the views

Reply:
Turns out we already had a method for this on `Document` but it was totally unused outside of tests, and I agree it should be on the manager.